File: src/main/java/cspfj/StatisticsManager.java
package cspfj;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.collect.Maps.newHashMap;

import java.lang.reflect.Field;
import java.lang.reflect.Modifier;
import java.util.Arrays;
import java.util.Map;
import java.util.Map.Entry;
import java.util.SortedMap;
import java.util.logging.Logger;

import com.google.common.base.Function;
import com.google.common.base.Predicate;
import com.google.common.base.Predicates;
import com.google.common.collect.BiMap;
import com.google.common.collect.HashBiMap;
import com.google.common.collect.HashMultimap;
import com.google.common.collect.Iterables;
import com.google.common.collect.Maps;
import com.google.common.collect.Multimap;

public final class StatisticsManager {

    private static final Map<String, Field> STATIC_STATISTICS = newHashMap();

    private static final BiMap<Object, String> REFD_OBJECTS = HashBiMap
            .create();

    private static final Multimap<Object, Field> DYNAMIC_FIELDS = HashMultimap
            .create();

    private static final Predicate<Field> ANNOTED_FIELD = new Predicate<Field>() {
        @Override
        public boolean apply(final Field input) {
            return input.getAnnotation(cspfj.util.Statistic.class) != null;
        }
    };
    private static final Predicate<Field> STATIC_FIELD = new Predicate<Field>() {
        @Override
        public boolean apply(final Field input) {
            return (input.getModifiers() & Modifier.STATIC) != 0;
        }
    };
    private static final Function<Field, Object> STATIC_FIELD_TO_VALUE = new Function<Field, Object>() {
        @Override
        public Object apply(final Field value) {
            try {
                return value.get(null);
            } catch (IllegalAccessException e) {
                throw new IllegalStateException(e);
            }
        }
    };

    private static final Logger LOGGER = Logger
            .getLogger(StatisticsManager.class.getName());

    private StatisticsManager() {

    }

    public static void register(final Class<?> clazz) {
        for (Field f : Iterables.filter(
                Arrays.asList(clazz.getDeclaredFields()),
                Predicates.and(STATIC_FIELD, ANNOTED_FIELD))) {
            STATIC_STATISTICS.put(clazz.getName() + '.' + f.getName(), f);
            f.setAccessible(true);
        }
    }

    public static void register(final String name, final Object object) {
        final Object oldObject = REFD_OBJECTS.inverse().get(name);
        if (oldObject != null) {
            LOGGER.info(name
                    + ": an object with the same name is already registered");
            DYNAMIC_FIELDS.removeAll(oldObject);
            REFD_OBJECTS.remove(oldObject);
        }

        REFD_OBJECTS.put(object, name);
        final Class<?> clazz = object.getClass();
        for (Field f : Iterables.filter(
                Arrays.asList(clazz.getDeclaredFields()),
                Predicates.and(ANNOTED_FIELD, Predicates.not(STATIC_FIELD)))) {
            DYNAMIC_FIELDS.put(object, f);
        }
    }

    public static Object get(final String name) {
        final Field staticField = STATIC_STATISTICS.get(name);
        if (staticField == null) {
            final int fieldNameAt = name.lastIndexOf('.');
            final Object obj = checkNotNull(REFD_OBJECTS.inverse().get(
                    name.substring(0, fieldNameAt)));
            final String fieldName = name.substring(fieldNameAt + 1,
                    name.length());
            for (Field f : DYNAMIC_FIELDS.get(obj)) {
                if (f.getName().equals(fieldName)) {
                    try {
                        return f.get(obj);
                    } catch (IllegalAccessException e) {
                        throw new IllegalStateException(e);
                    }
                }
            }
            throw new IllegalArgumentException("Could not find " + name);
        }
        return STATIC_FIELD_TO_VALUE.apply(staticField);

    }

    public static Map<String, Object> digest() {
        final SortedMap<String, Object> digest = Maps.newTreeMap();
        digest.putAll(Maps.transformValues(STATIC_STATISTICS,
                STATIC_FIELD_TO_VALUE));

        for (Entry<Object, Field> e : DYNAMIC_FIELDS.entries()) {
            try {
                digest.put(REFD_OBJECTS.get(e.getKey()) + "."
                        + e.getValue().getName(), e.getValue().get(e.getKey()));
            } catch (IllegalAccessException e1) {
                throw new IllegalStateException(e1);
            }
        }
        return digest;
    }

    private static boolean isIntType(final Class<?> input) {
        return Integer.class.equals(input) || int.class.equals(input)
                || Long.class.equals(input) || long.class.equals(input);
    }

    private static boolean isFloatType(final Class<?> input) {
        return Float.class.equals(input) || float.class.equals(input)
                || Double.class.equals(input) || double.class.equals(input);
    }

    public static void reset() {
        REFD_OBJECTS.clear();
        DYNAMIC_FIELDS.clear();
        for (Field f : STATIC_STATISTICS.values()) {
            try {
                if (isIntType(f.getType())) {
                    f.set(null, 0);
                } else if (isFloatType(f.getType())) {
                    f.set(null, 0f);
                }
            } catch (IllegalAccessException e) {
                throw new IllegalStateException(e);
            }
        }
    }

}

File: src/main/scala/cspfj/StatisticsManager.scala
package cspfj;

import scala.collection.mutable.MultiMap
import scala.collection.mutable.HashMap
import cspfj.util.Loggable
import java.lang.reflect.Modifier
import java.lang.reflect.Field
import scala.annotation.tailrec

object StatisticsManager extends Loggable {

  var objects: Map[String, AnyRef] = Map.empty

  def register(name: String, o: AnyRef) {
    require(!o.isInstanceOf[Class[_]])
    if (objects.contains(name)) {
      info(name + ": an object with the same name is already registered");
    }

    objects += name -> o
    //    for (
    //      f <- o.getClass.getDeclaredFields if annotedInstanceVariable(f)
    //    ) {
    //      f.setAccessible(true);
    //    }
  }

  private def annoted(f: Field) =
    f.getAnnotation(classOf[cspfj.Statistic]) != null
  //&&
  //(f.getModifiers & Modifier.STATIC) == 0

  def get(name: String) = {

    val fieldNameAt = name.lastIndexOf('.')
    val obj = objects.get(name.substring(0, fieldNameAt)).get
    val fieldName = name.substring(fieldNameAt + 1, name.length)
    obj.getClass.getDeclaredFields.find(f => annoted(f) && f.getName == fieldName) match {
      case Some(f) => { f.setAccessible(true); f.get(obj) }
      case None => throw new IllegalArgumentException("Could not find " + name + " (" + fieldName + " in " + obj.getClass.getDeclaredFields.toList + ")")
    }

  }

  def digest = (objects map {
    case (s, o) =>
      o.getClass.getDeclaredFields.filter(annoted).map { f =>
        f.setAccessible(true)
        (s + "." + f.getName) -> f.get(o)
      }
  } flatten).toMap

  def display = digest.map(t => t._1 + " = " + t._2).toSeq.sorted.mkString("\n")

  def isIntType(input: Class[_]) =
    input == classOf[Int] || input == classOf[Long]

  def isFloatType(input: Class[_]) = input == classOf[Float] || input == classOf[Double]

  def reset() {
    //static = Map.empty
    objects = Map.empty

  }

  def average[A](s: Seq[A])(implicit n: Numeric[A]) = n.toDouble(s.sum) / s.size

  def stDev[A](s: Seq[A])(implicit n: Numeric[A]) = {
    val avg = average(s)
    val sumSq = s map (v => math.pow(n.toDouble(v) - avg, 2)) sum

    math.sqrt(sumSq / (s.size - 1))
  }

  @tailrec
  def findKMedian[A](arr: Seq[A], k: Int, o: Ordering[A]): A = {
    val pivot = arr(scala.util.Random.nextInt(arr.size))
    val (s, b) = arr partition (o.gt(pivot, _))
    if (s.size == k) pivot
    // The following test is used to avoid infinite repetition
    else if (s.isEmpty) {
      val (s, b) = arr partition (pivot ==)
      if (s.size > k) pivot
      else findKMedian(b, k - s.size, o)
    } else if (s.size < k) findKMedian(b, k - s.size, o)
    else findKMedian(s, k, o)
  }

  def median[A](arr: Seq[A])(implicit o: Ordering[A]) =
    findKMedian(arr, arr.size / 2, o)

  def time[A](f: => A) = {
    var t = -System.currentTimeMillis
    val r = f
    t += System.currentTimeMillis()
    (r, t / 1000.0)
  }

}
